Introduce string interpolation support

Mats-SX opened the pull request:
See the CIP for details: https://docs.google.com/document/d/1hq-Pu7igb1aFdFZtVCnG59D1_QuqnzsBLPYm7dqpKwg/edit

Mats-SX committed:
Add basic support for interpolated strings to Cypher

o Parsing
o Rewriting of interpolation literals to interpolations
o Constructing command-level objects and implement their runtime evaluation

Conflicts:
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ResultIterator.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/conditions/collectNodesOfType.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/conditions/noDuplicatesInReturnItems.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ToStringFunction.scala
	community/cypher/frontend-2.3/src/main/scala/org/neo4j/cypher/internal/frontend/v2_3/parser/Base.scala

pontusmelke commented on community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/Interpolation.scala:
```diff
@@ -0,0 +1,79 @@
+    remaining: NonEmptyList[Either[Expression, String]],
+    ctx: ExecutionContext,
+    state: QueryState,
+    acc: mutable.Builder[Either[String, String], Option[NonEmptyList[Either[String, String]]]])
```
Either[String, String] looks a bit weird, could it not just build up a `NonEmptyList[String]`? 
`InterpolationValue` doesn't seem to care if the string is coming from an expression or a String?

boggle replied:
It will make a difference for later use by LIKE

pontusmelke replied:
Could we then have different case classes for the different cases instead, so you don't need to keep track of that `Left` means _from expression_ and `Right` means _from String_. So something like, `NonEmptyList[InterpolatedString]`, and have different implementations for the different cases?

pontusmelke commented on community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/values/InterpolationValue.scala:
```diff
@@ -0,0 +1,34 @@
+
+import org.neo4j.cypher.internal.compiler.v2_3.helpers.NonEmptyList
+
+case class InterpolationValue(parts: NonEmptyList[Either[String, String]]) {
```
As previous comment, couldn't `parts` just be an `NonEmptyList[String]`

Mats-SX replied:
With the above idea of different case classes, I tend to agree.

pontusmelke commented:
If we have not completely given up on the `compatibility-suite`, some or all of the acceptance tests should probably live there instead.

Mats-SX committed:
Add interpolation support for string functions

Conflicts:
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/Interpolation.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/StringFunctions.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/values/InterpolationValue.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/LTrim.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Left.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Lower.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/RTrim.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Right.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Split.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Substring.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Upper.scala

pontusmelke commented on community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/conditions/collectNodesOfType.scala:
```diff
@@ -23,10 +23,19 @@ import org.neo4j.cypher.internal.compiler.v2_3.ast.ASTNode
 import scala.reflect.ClassTag
 
-case class collectNodesOfType[T <: ASTNode](implicit tag: ClassTag[T]) extends (Any => Seq[T]) {
+case class collectNodesOfType[T <: ASTNode](include: Option[T => Boolean] = None)(implicit tag: ClassTag[T]) extends (Any => Seq[T]) {
```
Could we have `include: T => Boolean = _ => true` and then just do

```
case node: ASTNode if tag.runtimeClass.isAssignableFrom(node.getClass) && include(node) =>
    (acc) => acc :+ node.asInstanceOf[T]
```

?

boggle committed:
Add interpolation support for dynamic property lookup

o Rename InterpolationString => InterpolationStringPart

Conflicts:
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/values/InterpolationValue.scala

boggle committed:
Add interpolation support for regular expressions

o Rename regular expression objects to use "MatchRegex" prefix
o Literal parts of interpolations are treated as constant text to be matched

Conflicts:
	community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/RegularExpressionPredicateTest.scala

boggle committed:
Introduce InterpolationMode and some cleanup

Conflicts:
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ResultIterator.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/Interpolation.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/StringFunctions.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/ToStringFunction.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/predicates/MatchRegex.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/values/InterpolationValue.scala
	community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/MatchRegexTest.scala
	community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/InterpolationValueTest.scala

boggle committed:
Add interpolation support for LIKE

o Includes interpolation support for index seek by prefix

Conflicts:
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/SeekRange.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/convert/commands/ExpressionConverters.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/indexQuery.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/predicates/MatchRegex.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/values/InterpolationValue.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/pipes/NodeIndexSeekPipe.scala
	community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/plans/Sargable.scala

pontusmelke committed:
Add interpolation support for + operator

pontusmelke committed:
Add interpolation support for =, <>, <, <=, >, >=

pontusmelke committed:
Add interpolation support for aggregation keys

pontusmelke committed:
Add interpolations support for property updates

pontusmelke committed:
Add more tests and some clean up

pontusmelke committed:
Add interpolation support for legacy index

pontusmelke committed:
Add interpolation support for index range queries

pontusmelke committed:
Add interpolation support for equality and index lookup
